kcp/controllers/kyma_controller.go

```diff
@@ -0,0 +1,160 @@
+			return ctrl.Result{}, nil //nolint:wrapcheck
+		}
+		return ctrl.Result{}, err //nolint:wrapcheck
+	}
+
+	// Kyma resource was created or updated
+	// Get installed modules from Kyma CR
+	modules := getModulesList(kymaCR)
+	if len(modules) == 0 {
+		return ctrl.Result{}, errors.New("module list of KymaCR is empty")
+	}
+	for _, module := range modules {
+		watcherCR, err := r.getWatcherCR(ctx, module, kymaCR.Namespace)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if value, ok := watcherCR.Labels[DefaultOperatorWatcherCRLabel]; ok && strings.ToLower(value) == "true" {
+			watcherConfigMap, err := r.getWatcherCM(ctx, module, kymaCR.Namespace)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			err = r.updateConfigMap(ctx, watcherConfigMap, kymaCR)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			logger.Info("Corresponding ConfigMap of WatcherCR got updated.")
+			return ctrl.Result{}, nil
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KymaReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	SchemeBuilder := &scheme.Builder{GroupVersion: kyma.GroupVersion}
+	SchemeBuilder.Register(&kyma.Kyma{})
+	// Create ControllerBuilder
+	controllerBuilder := ctrl.NewControllerManagedBy(mgr).
+		For(
+			&kyma.Kyma{},
+			builder.WithPredicates(predicate.GenerationChangedPredicate{}),
+		)
+
+	return controllerBuilder.Complete(r)
+}
+
+func getModulesList(kymaResource *kyma.Kyma) []kyma.Module {
+	modules := kymaResource.Spec.Modules
+	return modules
+}
+
+func (r *KymaReconciler) getWatcherCR(ctx context.Context,
+	module kyma.Module,
+	namespace string,
+) (*componentv1alpha1.Watcher, error) {
+	watcherCRName := fmt.Sprintf("%s-%s", module.Name, module.Channel)
+	watcherCR := &componentv1alpha1.Watcher{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: namespace, Name: watcherCRName}, watcherCR); err != nil {
+		return nil, err
+	}
+	return watcherCR, nil
+}
+
+func (r *KymaReconciler) getWatcherCM(ctx context.Context,
+	module kyma.Module,
+	namespace string,
+) (*v1.ConfigMap, error) {
+	watcherConfigMapName := fmt.Sprintf("%s-%s", module.Name, module.Channel)
+	watcherConfigMap := &v1.ConfigMap{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: namespace, Name: watcherConfigMapName},
+		watcherConfigMap); err != nil {
+		return nil, err
+	}
+	return watcherConfigMap, nil
+}
+
+func (r *KymaReconciler) updateConfigMap(ctx context.Context,
+	watcherConfigMap *v1.ConfigMap,
+	kymaCR *kyma.Kyma,
+) error {
+	if watcherConfigMap.Data == nil {
+		// initialize data map, if map is nil
+		watcherConfigMap.Data = make(map[string]string)
+	} else {
+		for key, value := range watcherConfigMap.Data {
+			if key == kymaCR.Name && value == kymaCR.Namespace {
+				// Kyma already exists in ConfigMap, nothing has to be done
+				return nil
+			}
+		}
+	}
+	// Kyma does not exists in ConfigMap
+	watcherConfigMap.Data[kymaCR.Name] = kymaCR.Namespace
```

Comment:
this is not the intended use (end result) right? I thought config map will store a module list for all kymas?

Reply:
Lets have a small discussion round regarding this. I had in mind, that we a one ConfigMap for each WatcherCR containing the KymaCRs the Watcher needs to take care of. Thus, we do not have too many racing conditions when updating the ConfigMap.